src/dask_awkward/lib/optimize.py

```diff
@@ -474,5 +474,7 @@ def _prune_wildcards(columns: list[str], meta: AwkwardArray) -> list[str]:
         else:
             if imeta.fields:
                 good_columns.append(f"{col[:-2]}.{imeta.fields[0]}")
+            else:
+                good_columns.append(col[:-2])
```

Comment:
@douglasdavis 

As I understand it, this pruning logic works at the "column" level, rather than buffer level. 

For completeness, a restriction of this approach is that it would involve over-reading in the case that the underlying data are stored as separate buffers, i.e. `array.offsets` can be read separately from `array.content`. I vaguely recall discussing this in a meeting a while ago. Does anyone else recall if we touched upon this?

This patch ensures that if a column has _no_ child fields, we include the column. Are you happy with this change? I don't want to make a mis-step here.

Reply:
> As I understand it, this pruning logic works at the "column" level, rather than buffer level. Is this a fair description? I.e., we work at the buffer level (form keys) in Awkward, but in this library we speak about columns (which may be composites of buffers).

Yes! this code is definitely written around the idea of columns

> If so, a restriction of this approach is that it would lead to over-reading in the case that the underlying data are stored as separate buffers, i.e. if array.offsets can be read separately from array.content. I vaguely recall discussing this in a meeting a while ago. Does anyone else recall if we touched upon this?

I _kind of_ recall this as well, but I don't recall us discussing any further than what you've mentioned (i.e. the conversation was pretty much "it could happen" and not a discussion about solutions/workarounds)

> This patch ensures that if a column has no child fields, we just include the column itself. Are you happy with this change? I don't want to make a mis-step, but that seems like the appropriate fall-back to generalise this logic to non-record arrays.

> The consequence of this change is that the reproducer in this PR description no longer fails because we must read Muon_phi. That is to say, the expected behavior is that Muon_phi be read, because we need its offsets (which can only be read if the column itself is entirely read).

I need to re familiarize myself with this code a bit before concocting a reasonable answer!

Reply:
@agoose77 yes - that's what we do for most nanoevents. We have some pieces of our DSL for loading bits of columns, offsets, content, etc. We take apart each `Muon_*` and index it by `nMuon`, for instance.

Reply:
Going back to @agoose77's comment:

> This patch ensures that if a column has no child fields, we just include the column itself. Are you happy with this change? I don't want to make a mis-step, but that seems like the appropriate fall-back to generalise this logic to non-record arrays.

> The consequence of this change is that the reproducer in this PR description no longer fails because we must read Muon_phi. That is to say, the expected behavior is that Muon_phi be read, because we need its offsets (which can only be read if the column itself is entirely read).

> I assume that to avoid the read of Muon_phi, the proper solution is to write an interface to the IO layer that can encode the appropriate prior (that Muon_ branches all share offsets). I believe this is what the remapping logic by @lgray does, right?

At this point I'm happy with the change if it fixes something and we're still able to avoid unnecessary reads (for example through the mechanism @lgray has with remapping).

I am just generally curious how you came across that triggered the need to add the
```python
else:
    good_columns.append(col[:-2])
```
branch

Reply:
I'll give it a try properly early next week and report back!

Reply:
> I am just generally curious how you came across whatever triggered the need to add the

Just saw this - At the time I'd been playing around with getting uproot working.

IIRC I was following this optimisation through, and encountered a situation where a dataless offsets buffer appeared alongside a non-dataless offsets buffer in a context that demanded both arrays not be dataless. I think it was broadcasting. Tracking this back, I could see that the wildcard `data_touched` entry was not being considered, which led me to this statement.